proptest-state-machine/src/iterative_runner.rs

```diff
@@ -0,0 +1,861 @@
+use std::{cell::RefCell, marker::PhantomData};
+
+use proptest::{
+    bits::{BitSetLike, VarBitSet},
+    collection::SizeRange,
+    num::sample_uniform_incl,
+    prelude::Just,
+    std_facade::fmt::Debug,
+    strategy::{BoxedStrategy, Strategy, ValueTree},
+    test_runner::{Config, Reason, TestError, TestRunner},
+};
+
+pub fn test_sequential<
+    T,
+    MS,
+    SS,
+    M: IterativeModelStateMachine<ModelState = MS, Transition = T, SutState = SS>,
+    S: IterativeSutStateMachine<SutState = SS, Transition = T, ModelState = MS>,
+>(
+    config: Config,
+    mut mod_state: MS,
+    mut sut_state: SS,
+    transitions: Vec<T>,
+) where
+    T: Clone + Debug,
+{
+    let trans_len = transitions.len();
+    #[cfg(feature = "std")]
+    if config.verbose >= super::INFO_LOG {
+        eprintln!();
+        eprintln!("Running a test case with {} transitions.", trans_len);
+    }
+    #[cfg(not(feature = "std"))]
+    let _ = (config, trans_len);
+
+    // Check the invariants on the initial state
+    S::check_invariants(&sut_state, &mod_state);
+
+    for (ix, transition) in transitions.into_iter().enumerate() {
+        #[cfg(feature = "std")]
+        if config.verbose >= super::INFO_LOG {
+            eprintln!();
+            eprintln!(
+                "Applying transition {}/{}: {:?}",
+                ix + 1,
+                trans_len,
+                transition
+            );
+        }
+        #[cfg(not(feature = "std"))]
+        let _ = ix;
+        assert_eq!(
+            true,
+            M::preconditions(&mod_state, &sut_state, &transition),
+            "precondition failed on transition: {:?}",
+            transition
+        );
+        // Apply the transition on the states
+        mod_state = M::apply(mod_state, &transition);
+        sut_state = S::apply(sut_state, &mod_state, transition);
+
+        // Check the invariants after the transition is applied
+        S::check_invariants(&sut_state, &mod_state);
+    }
+
+    S::teardown(sut_state, mod_state)
+}
+
+pub trait IterativeModelStateMachine {
+    /// Model state used to drive the generation of the transitions
+    type ModelState: Clone + Debug;
+    /// Transition generated by this model
+    type Transition: Clone + Debug;
+    /// Concrete SUT state, which could be consulted to further sharpern
+    /// the state transition generation
+    type SutState;
+
+    /// Initial state of the model
+    fn init_state() -> BoxedStrategy<Self::ModelState>;
+
+    /// Generate transition based on the state of the model. System under
+    /// test can be consulted here as well to reduce number of generated
+    /// transitions
+    fn transition(
+        mod_state: &Self::ModelState,
+        sut_state: &Self::SutState,
+    ) -> BoxedStrategy<Self::Transition>;
```

Comment:
ModelState and SutState may diverge, and what is more important and differs this from the intent of SequentialRuner is that ModelState here is a supporting entity, rather than the reference model. 

Transition is indeed infallible, and perhaps the better way would be to treat that in the same way as failure to apply a transition to the model or to the system under test.